File: frontend/demo/component/side-nav/side-nav-hierarchy.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/side-nav';
import '@vaadin/icon';
import '@vaadin/icons';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('side-nav-hierarchy')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  protected override render() {
    return html`
      <div class="side-nav-sample">
        <div>
          <!-- tag::snippet[] -->
          <vaadin-side-nav style="width:100%">
            <vaadin-side-nav-item path="/messages">
              <vaadin-icon icon="vaadin:envelope" slot="prefix"></vaadin-icon>
              Messages
              <vaadin-side-nav-item path="/inbox" slot="children">
                <vaadin-icon icon="vaadin:inbox" slot="prefix"></vaadin-icon>
                Inbox
              </vaadin-side-nav-item>
              <vaadin-side-nav-item path="/sent" slot="children">
                <vaadin-icon icon="vaadin:paperplane" slot="prefix"></vaadin-icon>
                Sent
              </vaadin-side-nav-item>
              <vaadin-side-nav-item path="/trash" slot="children">
                <vaadin-icon icon="vaadin:trash" slot="prefix"></vaadin-icon>
                Trash
              </vaadin-side-nav-item>
            </vaadin-side-nav-item>
            <vaadin-side-nav-item>
              <vaadin-icon icon="vaadin:cog" slot="prefix"></vaadin-icon>
              Admin
              <vaadin-side-nav-item path="/users" slot="children">
                <vaadin-icon icon="vaadin:group" slot="prefix"></vaadin-icon>
                Users
              </vaadin-side-nav-item>
              <vaadin-side-nav-item path="/permissions" slot="children">
                <vaadin-icon icon="vaadin:key" slot="prefix"></vaadin-icon>
                Permissions
              </vaadin-side-nav-item>
            </vaadin-side-nav-item>
          </vaadin-side-nav>
          <!-- end::snippet[] -->
        </div>
      </div>
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/sidenav/SideNavHierarchy.java
package com.vaadin.demo.component.sidenav;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.sidenav.SideNav;
import com.vaadin.flow.component.sidenav.SideNavItem;
import com.vaadin.flow.router.Route;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.component.sidenav.views.*;

@Route("side-nav-hierarchy")
public class SideNavHierarchy extends Div {

    public SideNavHierarchy() {
        // tag::snippet[]
        SideNav nav = new SideNav();

        SideNavItem messagesLink = new SideNavItem("Messages", MessagesView.class, VaadinIcon.ENVELOPE.create());
        messagesLink.addItem(new SideNavItem("Inbox", InboxView.class, VaadinIcon.INBOX.create()));
        messagesLink.addItem(new SideNavItem("Sent", SentView.class, VaadinIcon.PAPERPLANE.create()));
        messagesLink.addItem(new SideNavItem("Trash", TrashView.class, VaadinIcon.TRASH.create()));
        
        SideNavItem adminSection = new SideNavItem("Admin");
        adminSection.setPrefixComponent(VaadinIcon.COG.create());
        adminSection.addItem(new SideNavItem("Users", UsersView.class, VaadinIcon.GROUP.create()));
        adminSection.addItem(new SideNavItem("Permissions", PermissionsView.class, VaadinIcon.KEY.create()));

        nav.addItem(messagesLink, adminSection);
        // end::snippet[]

        Div navWrapper = new Div(nav);
        nav.setWidthFull();
        add(navWrapper);
        
        this.addClassName("side-nav-sample");
    }

    public static class Exporter extends DemoExporter<SideNavHierarchy> { // hidden-source-line
    } // hidden-source-line
}
